torch/profiler/_pattern_matcher.py

```diff
@@ -345,6 +345,58 @@ def match(self, event: _ProfilerEvent):
         return False
 
 
+class SynchronizedDataLoaderPattern(Pattern):
+    '''
+    This pattern identifies if we are using num_workers=0 in DataLoader.
+    example:
+    torch.utils.data.DataLoader(dataset, batch_size=batch_size)
+    Add num_workers=N to the arguments. N depends on system configuration.
+
+    Pattern:
+    dataloader.py(...): __iter__
+        dataloader.py(...): _get_iterator
+            NOT dataloader.py(...): check_worker_number_rationality
+
+    Algorithm:
+    If we don't see check_worker_number_rationality call in the dataloader __iter__,
+    It is not an asynchronous dataloader.
+
+    '''
+
+    def __init__(self, prof: profile):
+        super().__init__(prof)
+        self.name = "Synchronized DataLoader Pattern"
+        self.description = (
+            "Detected DataLoader running with synchronized implementation. "
+            "Please enable asynchronous dataloading by setting num_workers > 0 when initializing DataLoader."
+        )
+        # We precompile the regex to avoid recompiling it every time we match
+        # The four back slashes are to escape windows path separator
+        seperator = "\\\\" if os.sep == "\\" else os.sep
+        self.iter_regex = re.compile(
+            r"torch/utils/data/dataloader.py\([0-9]+\): __iter__$"
+            .replace("/", seperator))
+        self.get_iterator_regex = re.compile(
+            r"torch/utils/data/dataloader.py\([0-9]+\): _get_iterator$"
+            .replace("/", seperator))
+        self.check_worker_regex = re.compile(
+            r"torch/utils/data/dataloader.py\([0-9]+\): check_worker_number_rationality$"
+            .replace("/", seperator))
+
+    def match(self, event: _ProfilerEvent):
```

Comment:
This could use some comments to explain your reasoning. I think it also needs some sort of check for "waiting". If my dataloader is able to keep up with training (maybe I already have the data in memory, maybe I'm handling parallelism elsewhere, maybe it's just really cheap) then `num_workers=0` is appropriate.

Reply:
(You'll have to override this for test stability, of course.)

Reply:
I think I should add a TODO here. I was planning to add a check to the idle time during data loading, but I want to do it after kineto type is in.